### PR TITLE
Issue 14147 - Compiler crash on identical functions in a single module

### PR DESCRIPTION
**Pull request by 9rnsr**
https://issues.dlang.org/show_bug.cgi?id=14147

Partial fix for [issue 2789](https://issues.dlang.org/show_bug.cgi?id=2789).

### CODE REVIEW

**Comment by 9rnsr** [comment]
Ready to merge.

File: src/func.c [hunk 1]
```diff
@@ -1222,6 +1223,49 @@ void FuncDeclaration::semantic(Scope *sc)
...
+                    fd->error("%s conflicts with at %s",
+                            parametersTypeToChars(tf2->parameters, tf2->varargs), loc.toChars());
+                    fd->type = Type::terror;
+                    fd->errors = true;
+                }
+            }
+            else if (AliasDeclaration *ad = s->isAliasDeclaration())
+            {
+                // todo
+                s = ad->overnext;
```
Comment (WalterBright) [inline src/func.c]: I suspect much of this function is duplicating what is in overloadApply(), but not handling all the cases.

File: src/mangle.c [hunk 2]
```diff
@@ -901,3 +901,26 @@ void mangleToBuffer(Expression *e, OutBuffer *buf)
     Mangler v(buf);
     e->accept(&v);
 }
+
+void mangleToFuncSignature(OutBuffer *buf, FuncDeclaration *fd)
```
Comment (WalterBright) [inline src/mangle.c]: I don't understand why there is a same-only-different type signature routine. Why should this produce different results than tf->deco? And if it does produce something different, isn't something terribly wrong somewhere?

File: src/func.c [hunk 3]
```diff
@@ -1222,6 +1223,49 @@ void FuncDeclaration::semantic(Scope *sc)
...
+                s = fd->overnext;
+                if (fd->errors)
+                    continue;
+
+                // Don't have to check conflict between declaration and definition.
+                if ((fbody != NULL) != (fd->fbody != NULL))
+                    continue;
+
+                buf2.reset();
+                mangleToFuncSignature(&buf2, fd);
```
Comment (WalterBright) [inline src/func.c]: By computing the signature over and over, this can result in terrible compiler performance.

The  type `deco` system, on the other hand, produces a signature only once and uses a hash for lookup.

**Comment by yebblies** [comment]
I'm not sure the mangling approach is correct.  For example, this should be an error:
```
extern(D) void func() {}
extern(C) void func() {}
```

Is there some reason the check for matching (argument) signatures can't be done on `overloadInsert`?  Why is `semantic2` an appropriate place for this?

**Comment by 9rnsr** [comment]
`overloadInsert` is usually called from `addMember`. In that time, we cannot invoke semantic analysis to see correct function signatures because forward reference resolution is not yet ready (sibling scope members' `setScope` are not yet called).

In `semantic2`, function parameter types should be analyzed already in `semantic`. So it's the best place to see conflicts.

**Comment by yebblies** [comment]
Is that for global functions too or just member functions?

Could it be done at the end of the parent symbol's semantic1 routine?

**Comment by 9rnsr** [comment]
> Is that for global functions too or just member functions?

For all functions in overloadable scopes. So includIng global functions.

> Could it be done at the end of the parent symbol's semantic1 routine?

Possible, but it needs modifying many places - Module.semantic, StructDeclaration.semantic, TemplateInstance.semantic, ... Instead do it in FuncDeclatation.semantic2 is much more sImple.

File: src/func.c (hunk 1 above)
Comment (9rnsr) [inline src/func.c]: Now `overloadApply` is used to iterate function overloads.

File: src/mangle.c (hunk 2 above)
Comment (9rnsr) [inline src/mangle.c]: As I commented to, overload conflict check does not need function return type. Even if two auto return functions defined in non-root module, their conflicts can be detected without their semantic3.

```d
auto foo() { ... }
auto foo() { ... }   // compiler can detect conflict, without semantic3 running
```

File: src/func.c (hunk 3 above)
Comment (9rnsr) [inline src/func.c]: As I already mentioned, `FuncDeclaratio.type.deco` is over-spec for conflict check. If we worry about performance, the `mangleToFuncSignature` result can be cached into `FuncDeclaration`.

**Comment by 9rnsr** [comment]
Rebased onto master, and add support to detect `extern (C)` function conflicts.

**Comment by yebblies** [comment]
Still doesn't error on this?
```
extern(D) void func() {}
extern(C) void func() {}
```

**Comment by 9rnsr** [comment]
> Still doesn't error on this?

No, now it will raise conflict error. See the case `f_MixExtern1` in `fail2789.d`.

File: src/dmangle.d [hunk 4]
```diff
@@ -862,3 +862,32 @@ extern (C++) void mangleToBuffer(Expression e, OutBuffer* buf)
...
+void mangleToFuncSignature(OutBuffer* buf, FuncDeclaration fd)
+{
+    assert(fd.type.ty == Tfunction);
+    auto tf = cast(TypeFunction)fd.type;
+
+    scope Mangler v = new Mangler(buf);
+
+    //if (fd.needThis() || fd.isNested())
+    //    buf.writeByte(Type.needThisPrefix());
+    MODtoDecoBuffer(buf, tf.mod);
```
Comment (yebblies) [inline src/dmangle.d]: Why is this necessary?  Which test cases are affected by this?

File: src/func.d [hunk 5]
```diff
@@ -1286,12 +1287,99 @@ public:
...
+        //printf("FuncDeclaration::semantic2 [%s] fd0 = %s %s\n", this.loc.toChars(), this.toChars(), this.type.toChars());
+
+        if (overnext && !errors)
+        {
+            OutBuffer buf1;
+            OutBuffer buf2;
+
+            // Always starts the lookup from 'this', because the conflicts with
+            // previous overloads are already reported.
+            auto f1 = this;
```
Comment (yebblies) [inline src/func.d]: Why have a copy in `f1`?  Wouldn't using `this` make the code easier to follow?

File: src/func.d [hunk 6]
```diff
@@ -1286,12 +1287,99 @@ public:
...
+                mangleToFuncSignature(&buf2, f2);
+
+                auto s1 = buf1.peekString();
+                auto s2 = buf2.peekString();
+
+                //printf("+%s\n\ts1 = %s\n\ts2 = %s @ [%s]\n", toChars(), s1, s2, f2.loc.toChars());
+                if (strcmp(s1, s2) == 0)
+                {
+                    auto tf2 = cast(TypeFunction)f2.type;
+                    f2.error("%s conflicts with at %s",
```
Comment (yebblies) [inline src/func.d]: This need a grammatical fix - maybe `"%s conflicts with %s at %s"` or `"%s conflicts with another declaration at %s"` if you want to avoid naming the second func.

File: src/func.d [hunk 7]
```diff
@@ -1286,12 +1287,99 @@ public:
...
+                    if (f == f2)
+                        return 0;
+                }
+
+                // extern (C) functions always conflict each other.
+                if (f1.ident == f2.ident &&
+                    f1.toParent2() == f2.toParent2() &&
+                    f1.linkage == LINKc && f2.linkage == LINKc)
+                {
+                    /* Allow the hack that is actually used in druntime,
```
Comment (yebblies) [inline src/func.d]: Could you please file a bug about this issue so we don't forget about it?

File: src/func.d [hunk 8]
```diff
@@ -1286,12 +1287,99 @@ public:
...
+                     *  alias sigfn_t2 = void function(int) nothrow @nogc;
+                     *  sigfn_t  bsd_signal(int sig, sigfn_t  func);
+                     *  sigfn_t2 bsd_signal(int sig, sigfn_t2 func) nothrow @nogc;  // no error
+                     */
+                    if (f1.fbody is null || f2.fbody is null)
+                        return 0;
+
+                    auto tf1 = cast(TypeFunction)f1.type;
+                    auto tf2 = cast(TypeFunction)f2.type;
+                    f2.error("%s conflicts with at %s because of extern (C)",
```
Comment (yebblies) [inline src/func.d]: `"extern (C) %s conflicts with another function at %s"`?
Or possibly
`"extern (C) functions cannot be overloaded"` and then list the conflicting overloads (with locations).  I think this would be best for avoiding confusion.

File: src/func.d (hunk 6 above)
Comment (yebblies) [inline src/func.d]: `"overloaded functions have the same argument lists"` and then list overloads?

**Comment by yebblies** [comment]
I think all linkages except C++ and D should not be able to be overloaded.

File: src/dmangle.d (hunk 4 above)
Comment (9rnsr) [inline src/dmangle.d]: `void foo()` and `void foo() const` can be overloaded each other.

File: src/func.d (hunk 5 above)
Comment (9rnsr) [inline src/func.d]: It's just for the code reading. To me `f1` vs `f2` is easier to understand than `this` vs `f2`.

File: src/func.d (hunk 7 above)
Comment (9rnsr) [inline src/func.d]: OK, will do.

File: src/func.d (hunk 8 above)
Comment (9rnsr) [inline src/func.d]: I'll update the message to: `"%s cannot be overloaded with another extern (C) function at %s"`.

File: src/func.d (hunk 6 above)
Comment (9rnsr) [inline src/func.d]: I'll update the message to `"%s conflicts with previous declaration at %s"`.

File: src/func.d (hunk 8 above)
Comment (yebblies) [inline src/func.d]: Great.  And please include `extern (Windows)` etc

File: src/func.d (hunk 6 above)
Comment (yebblies) [inline src/func.d]: Even better.

File: src/dmangle.d (hunk 4 above)
Comment (yebblies) [inline src/dmangle.d]: Huh, for some reason I thought this was only for non-member functions, and member functions were checked elsewhere.

File: src/func.d [hunk 9]
```diff
@@ -1286,12 +1287,99 @@ public:
...
+
+            // Always starts the lookup from 'this', because the conflicts with
+            // previous overloads are already reported.
+            auto f1 = this;
+            mangleToFuncSignature(&buf1, f1);
+
+            overloadApply(f1, (Dsymbol s)
+            {
+                auto f2 = s.isFuncDeclaration();
+                if (!f2|| f1 == f2 || f2.errors)
```
Comment (yebblies) [inline src/func.d]: Spacing around `||`

File: src/dmangle.d [hunk 10]
```diff
@@ -862,3 +862,32 @@ extern (C++) void mangleToBuffer(Expression e, OutBuffer* buf)
...
+ * if fd is an auto function.
+ */
+void mangleToFuncSignature(OutBuffer* buf, FuncDeclaration fd)
+{
+    assert(fd.type.ty == Tfunction);
+    auto tf = cast(TypeFunction)fd.type;
+
+    scope Mangler v = new Mangler(buf);
+
+    //if (fd.needThis() || fd.isNested())
```
Comment (yebblies) [inline src/dmangle.d]: Comments can be deleted now?

File: src/func.d (hunk 8 above)
Comment (9rnsr) [inline src/func.d]: `extern (Windows)` needs ... more work. If two `exntern (Windows) void foo(int*)` and `void foo(void*)` exist, they don't conflicts on semantic overload level, but they will have same link name `_foo@4` *only in 32bit codegen*.
On the other hand, in win64 `extern (Windows)` is equivalent with `extern (C)`.

Please give me some time...

File: src/func.d (hunk 9 above)
Comment (9rnsr) [inline src/func.d]: Will fix.

File: src/func.d (hunk 7 above)
Comment (9rnsr) [inline src/func.d]: Done. https://issues.dlang.org/show_bug.cgi?id=15217

**Comment by 9rnsr** [comment]
> I think all linkages except C++ and D should not be able to be overloaded.

I think one `extern(C)` and one or more `extern (D)` functions can be overloaded.

File: src/func.d (hunk 8 above)
Comment (yebblies) [inline src/func.d]: >`extern (Windows)` needs ... more work. If two `exntern (Windows) void foo(int*)` and `void foo(void*)` exist, they don't conflicts on semantic overload level, but they will have same link name `_foo@4` *only in 32bit codegen*.

I don't think it matters that they might have a different linker symbol name.  They're not designed for overloading, and I think it would be best to simply disallow it.

> Please give me some time...

Of course!

**Comment by yebblies** [comment]
> I think one `extern(C)` and one or more `extern (D)` functions can be overloaded.

Yes you're right.  I'm sure I've done that before.

File: src/func.d (hunk 8 above)
Comment (9rnsr) [inline src/func.d]: > They're not designed for overloading, and I think it would be best to simply disallow it.

OK, make sense.

**Comment by 9rnsr** [comment]
Now `extern (Windows)` and `extern (C)` conflict each other.

**Comment by CyberShadow** [comment]
>  Allow an overload declaration hack, which is actually used in druntime 

That use was added very recently, maybe we should fix Druntime instead?

**Comment by yebblies** [comment]
What about `extern (Pascal)`?

**Comment by 9rnsr** [comment]
I don't know well about `extern (Pascal)`.

**Comment by 9rnsr** [comment]
@CyberShadow I think, no.

**Comment by yebblies** [comment]
`extern (Pascal)` is similar to `extern (Windows)`, and does not support overloading.

**Comment by 9rnsr** [comment]
> does not support overloading

Where is it documented? I couldn't find any specification about `extern (Pascal)`, so I don't want to add any special behaviors for that.

**Comment by yebblies** [comment]
Pascal is an older variant of stdcall.

**Comment by 9rnsr** [comment]
OK, now `LINKc`, `LINKwindows`, `LINKpascal`, and `LINKobjc` conflict each other.

**Comment by yebblies** [comment]
> @CyberShadow I think, no.

Why shouldn't druntime be fixed?

**Comment by 9rnsr** [comment]
[Issue 15217](https://issues.dlang.org/show_bug.cgi?id=15217) looks intentional to propagate delegate argument attributes to the function call. Currently there's not alternative way to do it, so I think we should leave it.

**Comment by wilzbach** [comment]
Rebased to #7577
